util/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/AttributesUtil.java

```diff
@@ -0,0 +1,219 @@
+        || attrs.getStrings(Tag.ModalitiesInStudy).length <= 1) {
+      return new String[]{attributesToQidoPath(attrs, includeFields)};
+    }
+
+    List<String> result = new ArrayList<>();
+    for (String modality : attrs.getStrings(Tag.ModalitiesInStudy)) {
+      Attributes attrsCopy = new Attributes(attrs);
+      attrsCopy.setString(Tag.ModalitiesInStudy, VR.CS, modality);
+      result.add(attributesToQidoPath(attrsCopy, includeFields));
+    }
+    return result.toArray(new String[]{});
+  }
+
+  /**
+   * Returns corresponding QIDO-RS path
+   * @param attrs dcm4che Attributes to convert
+   * @param includeFields additonal includeFields for QIDO-RS
+   */
+  public static String attributesToQidoPath(Attributes attrs, String... includeFields)
+      throws DicomServiceException {
+    HashSet<Integer> nonEmptyKeys = new HashSet<>();
+    for (int tag : attrs.tags()) {
+      if (attrs.containsValue(tag)) {
+        nonEmptyKeys.add(tag);
+      }
+    }
+
+    StringBuilder qidoPath = new StringBuilder();
+    if (nonEmptyKeys.contains(Tag.QueryRetrieveLevel)) {
+      switch (attrs.getString(Tag.QueryRetrieveLevel)) {
+        case "PATIENT":
+          qidoPath.append("patients");
+          break;
+        case "STUDY":
+          qidoPath.append("studies");
+          break;
+        case "SERIES":
+          qidoPath.append("series");
+          break;
+        case "IMAGE":
+          qidoPath.append("instances");
+          break;
+        default:
+          throw new DicomServiceException(Status.ProcessingFailure,
+              "Invalid QueryRetrieveLevel specified");
+      }
+      nonEmptyKeys.remove(Tag.QueryRetrieveLevel);
+    } else {
+      throw new DicomServiceException(Status.ProcessingFailure, "No QueryRetrieveLevel specified");
+    }
+
+    if (nonEmptyKeys.size() > 0 || includeFields.length > 0) {
+      qidoPath.append("?");
+    }
+
+    if (includeFields.length > 0) {
+      for (String includeField : includeFields) {
+        qidoPath.append("includefield=" + includeField + "&");
+      }
+    }
+
+    for (int keyTag : nonEmptyKeys) {
+      // non-string type search keys don't seem to exist
+      // pass-through multiple values (valid for UID lists, ignored for rest)
+      for (String value : attrs.getStrings(keyTag)) {
+        String encodedValue;
+        try {
+          encodedValue = URLEncoder.encode(value, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+          throw new DicomServiceException(Status.ProcessingFailure, e);
+        }
+        qidoPath.append(TagUtils.toHexString(keyTag) + "=" + encodedValue + "&");
+      }
+    }
+
+    return qidoPath.toString();
+  }
+
+  /**
+   * Converts QIDO_RS json reply to dcm4che Attributes
+   * @param jsonDataset
+   */
+  public static Attributes jsonToAttributes(JSONObject jsonDataset) throws DicomServiceException {
+    Attributes attributes = new Attributes();
+    for (String tag : jsonDataset.keySet()) {
```

Comment:
The structure of PN VR attributes is different than regular strings (it can contain alphabetic, ideographic, and phonetic components).  Please see this example: http://dicom.nema.org/medical/dicom/current/output/html/part18.html#table_F.3.1-1

Reply:
Ok, added special handling for PN VR.